core/src/test/scala/org/labrad/data/DataBench.scala

```diff
@@ -34,67 +90,95 @@ object DataBench {
+    val doubleArrayData = Arr(doubleArray)
 
-    val unflatBE = TreeData.fromBytes(data.t, bytesBE)(BIG_ENDIAN)
-    val unflatLE = TreeData.fromBytes(data.t, bytesLE)(LITTLE_ENDIAN)
-
-    val flatBE = FlatData.fromBytes(data.t, bytesBE)(BIG_ENDIAN)
-    val flatLE = FlatData.fromBytes(data.t, bytesLE)(LITTLE_ENDIAN)
-
-    timeIt("flatten") {
-      for (i <- 0 until 1000) {
-        unflatBE.toBytes(BIG_ENDIAN)
-      }
-    }
-    timeIt("flattenFlat") {
-      for (i <- 0 until 1000) {
-        flatBE.toBytes(BIG_ENDIAN)
-      }
-    }
+    println("cluster array:")
+    bench(data1)
     println()
 
-    timeIt("flattenSwap") {
-      for (i <- 0 until 1000) {
-        unflatBE.toBytes(LITTLE_ENDIAN)
+    println("int array:")
+    bench(intArrayData)
+    benchGet("getFlatIter", intArrayData, data => {
+      data.flatIterator.map(_.get[Int]).toArray
+    })
+    benchGet("getFast", intArrayData, data => {
+      val bytes = data.arrayBytes
+      val n = data.arraySize
+      assert(bytes.length == 4*n)
+      val result = Array.ofDim[Int](n)
+      var i = 0
+      while (i < n) {
+        result(i) = bytes.getInt(4*i)
+        i += 1
       }
-    }
-    timeIt("flattenFlatSwap") {
-      for (i <- 0 until 1000) {
-        flatBE.toBytes(BIG_ENDIAN)
+      result
+    })
+    benchGet("get[Array[Int]]", intArrayData, _.get[Array[Int]])
+    benchSet("set:Arr(ints)", implicit endianness => {
+      Arr(intArray)
+    })
+    benchSet("set:fast", implicit endianness => {
+      val data = Data("*i")
+      val n = intArray.length
+      data.setArrayShape(n)
+      val bytes = data.arrayBytes
+      var i = 0
+      while (i < n) {
+        bytes.setInt(4 * i, intArray(i))
+        i += 1
       }
-    }
+      data
+    })
     println()
 
-    timeIt("unflattenBE") {
-      for (i <- 0 until 1000) {
-        TreeData.fromBytes(data.t, bytesBE)(BIG_ENDIAN)
+    println("double array:")
+    bench(doubleArrayData)
+    benchGet("getFlatIter", doubleArrayData, data => {
+      data.flatIterator.map(_.get[Double]).toArray
+    })
+    benchGet("getFast", doubleArrayData, data => {
+      val bytes = data.arrayBytes
+      val n = data.arraySize
+      assert(bytes.length == 8*n)
+      val result = Array.ofDim[Double](n)
+      var i = 0
+      while (i < n) {
+        result(i) = bytes.getDouble(8*i)
+        i += 1
       }
-    }
-    timeIt("unflattenFlatBE") {
-      for (i <- 0 until 1000) {
-        FlatData.fromBytes(data.t, bytesBE)(BIG_ENDIAN)
+      result
+    })
+    benchGet("get[Array[Double]]", doubleArrayData, _.get[Array[Double]])
+    benchSet("Arr(doubles)", implicit endianness => {
+      Arr(doubleArray)
+    })
+    benchSet("set:fast", implicit endianness => {
+      val data = Data("*v")
+      val n = doubleArray.length
+      data.setArrayShape(n)
+      val bytes = data.arrayBytes
+      var i = 0
+      while (i < n) {
+        bytes.setDouble(8 * i, doubleArray(i))
+        i += 1
```

Comment:
Alas, yes. `for` in scala is syntactic sugar calling various methods (`forEach`, `map`, `flatMap`, depending on the details of the loop) and the loop body gets turned into a closure that is passed into those methods. _Sometimes_ the compiler can optimize this, but otherwise you incur an extra function call on every loop iteration, which slows things down substantially. There are some macro packages that let you write for-loops that get macro-expanded into fast while loops, but I thought it'd be simplest to just write the loops directly here.